rust/Cargo.lock

```diff
@@ -431,1318 +831,3758 @@ dependencies = [
+name = "num-bigint"
+version = "0.4.5"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "c165a9ab64cf766f73521c0dd2cfdff64f488b8f0b3e621face3462d3db536d7"
+dependencies = [
+ "num-integer",
+ "num-traits",
+]
+
+[[package]]
+name = "num-bigint-dig"
+version = "0.8.4"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "dc84195820f291c7697304f3cbdadd1cb7199c0efc917ff5eafd71225c136151"
+dependencies = [
+ "byteorder",
+ "lazy_static",
+ "libm",
+ "num-integer",
+ "num-iter",
+ "num-traits",
+ "rand 0.8.5",
+ "smallvec",
+ "zeroize",
+]
+
+[[package]]
+name = "num-conv"
+version = "0.1.0"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "51d515d32fb182ee37cda2ccdcb92950d6a3c2893aa280e540671c2cd0f3b1d9"
+
+[[package]]
+name = "num-integer"
+version = "0.1.46"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "7969661fd2958a5cb096e56c8e1ad0444ac2bbcd0061bd28660485a44879858f"
+dependencies = [
+ "num-traits",
+]
+
+[[package]]
+name = "num-iter"
+version = "0.1.45"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "1429034a0490724d0075ebb2bc9e875d6503c3cf69e235a8941aa757d83ef5bf"
+dependencies = [
+ "autocfg",
+ "num-integer",
+ "num-traits",
+]
+
+[[package]]
+name = "num-traits"
+version = "0.2.19"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "071dfc062690e90b734c0b2273ce72ad0ffa95f0c74596bc250dcfd960262841"
+dependencies = [
+ "autocfg",
+ "libm",
+]
+
+[[package]]
+name = "num_cpus"
+version = "1.16.0"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "4161fcb6d602d4d2081af7c3a45852d875a03dd337a6bfdd6e06407b61342a43"
+dependencies = [
+ "hermit-abi 0.3.9",
+ "libc",
+]
+
+[[package]]
+name = "num_enum"
+version = "0.7.2"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "02339744ee7253741199f897151b38e72257d13802d4ee837285cc2990a90845"
+dependencies = [
+ "num_enum_derive",
+]
+
+[[package]]
+name = "num_enum_derive"
+version = "0.7.2"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "681030a937600a36906c185595136d26abfebb4aa9c65701cefcaf8578bb982b"
+dependencies = [
+ "proc-macro-crate 3.1.0",
+ "proc-macro2",
+ "quote",
+ "syn 2.0.66",
+]
+
+[[package]]
+name = "object"
+version = "0.35.0"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "b8ec7ab813848ba4522158d5517a6093db1ded27575b070f4177b8d12b41db5e"
+dependencies = [
+ "memchr",
+]
+
+[[package]]
+name = "once_cell"
+version = "1.19.0"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "3fdb12b2476b595f9358c5161aa467c2438859caa136dec86c26fdd2efe17b92"
+
+[[package]]
+name = "opaque-debug"
+version = "0.3.1"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "c08d65885ee38876c4f86fa503fb49d7b507c2b62552df7c70b2fce627e06381"
+
+[[package]]
+name = "openssl"
```

Comment:
Hmm. OpenSSL still has [frequent and often serious security vulnerabilities](https://openssl.org/news/vulnerabilities.html). These include some bugs that cause platform-dependent and context-dependent undefined behaviour, so won't necessarily show up in testing; for example, the bug with [not restoring XMM registers](https://openssl.org/news/secadv/20230908.txt) on x86-64 processors when using AVX512-IFMA instructions in the Poly1305 code last September. (As it happens that only affected Windows, but that's not the point.)

How is Arti depending on OpenSSL precisely? `tor-llcrypto` does not depend on OpenSSL. At first glance, it seems that [tor-rtcompat](https://crates.io/crates/tor-rtcompat) is depending on it via native-tls. But that should be an optional dependency (only needed for `tokio::TokioNativeTlsRuntime`) I think.

It appears that tor-rtcompat has a [native-tls](https://gitlab.torproject.org/tpo/core/arti/-/blob/main/crates/tor-rtcompat/Cargo.toml?ref_type=heads#L28) feature. Can we disable that?

Reply:
Arti does not; we can configure it to use Rustls / ring. Note however that this openssl dependency is only actually used on Linux; on iOS it will be using the built-in Security framework.

Reply:
I looked into this, and in order to keep the `ring` version unified, we would need to downgrade from `arti-client 0.11` (which is already a downgrade for `ring` unification) to `arti-client 0.9.0` (and pin a bunch of dependencies to specifically `0.9.0` etc, as Arti migrated to `rustls 0.21` in `0.9.1`).

I've pushed this in ce10808fa1475de0d289d723d9fe2ac2d69f7b02 for you to consider. It still needs testing to confirm it works.